query_test.go

```diff
@@ -0,0 +1,140 @@
+				return err
+			}
+		}
+		return nil
+	}
+
+	// Test an error is returned if the query is empty
+	_, err := s.client.EnSQL(context.Background(), &api.Query{})
+	require.ErrorIs(err, ensign.ErrEmptyQuery, "expected error for empty query")
+
+	// Test with a valid query
+	query := &api.Query{
+		Query: "SELECT * FROM topic",
+	}
+	cursor, err := s.client.EnSQL(context.Background(), query)
+	require.NoError(err, "expected no error for valid query")
+
+	// Fetch all the events at once
+	results, err := cursor.FetchAll()
+	require.NoError(err, "expected no error fetching all events")
+	require.Len(results, len(events), "expected all events to be returned")
+	results, err = cursor.FetchAll()
+	require.NoError(err, "expected no error fetching all events again")
+	require.Len(results, 0, "expected no more events to be returned")
+
+	// Fetch multiple events
+	cursor, err = s.client.EnSQL(context.Background(), query)
+	require.NoError(err, "expected no error for valid query")
+	results, err = cursor.FetchMany(2)
+	require.NoError(err, "expected no error fetching multiple events")
+	require.Len(results, 2, "expected two events to be returned")
+	results, err = cursor.FetchMany(2)
+	require.NoError(err, "expected no error fetching multiple events again")
+	require.Len(results, 1, "expected one event to be returned")
+	results, err = cursor.FetchMany(2)
+	require.NoError(err, "expected no error fetching multiple events again")
+	require.Len(results, 0, "expected no more events to be returned")
+
+	// Fetch one at a time
+	cursor, err = s.client.EnSQL(context.Background(), query)
+	require.NoError(err, "expected no error for valid query")
+	for i := 0; i < len(events); i++ {
+		event, err := cursor.FetchOne()
+		require.NoError(err, "expected no error fetching one event")
+		require.NotNil(event, "expected event to be returned")
+		require.Equal(events[i].Data, event.Data, "expected event data to match")
+
+		// Cannot ack or nack a query result
+		_, err = event.Ack()
+		require.ErrorIs(err, ensign.ErrCannotAck, "expected error; cannot ack query result")
+		_, err = event.Nack(api.Nack_UNKNOWN_TYPE)
+		require.ErrorIs(err, ensign.ErrCannotAck, "expected error; cannot nack query result")
+	}
+
+	// Should be no more events
+	event, err := cursor.FetchOne()
+	require.NoError(err, "expected no error fetching one event")
+	require.Nil(event, "expected no more events to be returned")
+
+	// After close the cursor returns an error
+	require.NoError(cursor.Close(), "expected no error closing cursor")
+	_, err = cursor.FetchOne()
+	require.ErrorIs(err, ensign.ErrCursorClosed, "expected error fetching one event after close")
+
+	// Test that an error is returned if the server returns an error
+	// TODO: The mock server is not returning an error, not sure why
+	//require.NoError(s.mock.UseError(mock.EnSQLRPC, codes.InvalidArgument, "unparseable query"))
+	//_, err = s.client.EnSQL(ctx, query)
+	//s.GRPCErrorIs(err, codes.InvalidArgument, "unparseable query")
```

Comment:
No idea why this test case isn't working. The mock function appears to be getting exercised, but the client is getting nil instead of the error.

Reply:
The error is doesn't come back from the stream until `Recv()` is called, so you will probably need to do a `FetchOne` to get the InvalidArgument error back. It's kind of annoying, but that's how streaming works unfortunately. This is one of the reasons we have stream initialization on publish and subscribe. We could think about doing something like that for Query too -- but perhaps a better idea is to have the cursor do a `FetchOne` immediately so that it can get the error and it has the first result cached? The `read` method could then return the value from the cache and fetch the next value and cache it.